# THIS FILE IS IN THE HANDOUTS FOLDER. COPY IT INTO YOUR CLASS NOTES

- [**Read the chapter on the website!**](https://ledatascifi.github.io/ledatascifi-2025/content/05/02_reg.html) It contains a lot of extra information we won't cover in class extensively.
- After reading that, I recommend [this webpage as a complimentary place to get additional intuition.](https://aeturrell.github.io/coding-for-economists/econmt-regression.html)

## Today

[Finish picking teams and declare initial project interests in the project sheet](https://docs.google.com/spreadsheets/d/1SMetWKgI3JdhFdBwihDgIY3BDubvihkJCWbgLcCe4fs/edit?usp=sharing)


# Today is mostly about INTERPRETING COEFFICIENTS (6.4 in the book)

1. 25 min reading groups: Talk/read through two regression pages (6.3 and 6.4) 
    - Assemble your own notes. Perhaps in the "Module 4 notes" file, but you can do this in any file you want.
    - After class, each group will email their notes to the TA/me for participation. (Effort grading.)
1. 10 min: class builds joint "big takeaways and nuanced observations" 
1. 5 min: Interpret models 1-2 as class as practice. 
1. 20 min reading groups: Work through remaining problems below.
1. 10 min: wrap up  

---

## Notes from 6.3 and 6.4

- 𝑅2=1−𝑆𝑆𝐸/𝑇𝑆𝑆
 = The fraction of variation in y that a model can explain, and always between 0 and 1.

- Adding extra variables to a model can not reduce R2
 even if the variables are random

- Adjusted R2 adjusts R2 by penalizing based on the number of variables in X. It can actually be negative.

-  sm_ols(<model and data>).fit().summary() automatically reports R2 and Adjusted R2

- Can have negative coefficients that are part of a model: When your regression omits a variable that is in the “true model” generating the data, it will bias the coefficients on other variables

## Table of how an increase in x (continuous) affects y

𝑦=𝑎+𝛽𝑋

If 𝑋↑
 1 unit, then 𝑦↑
 by 𝛽
 units

log𝑦=𝑎+𝛽𝑋

If 𝑋↑
 1 unit, then 𝑦↑
 by about 100∗𝛽
%.
(Exact: 100∗(exp(𝛽)−1)
)

𝑦=𝑎+𝛽log𝑋

If 𝑋↑
 1%, then 𝑦↑
 by about 𝛽/100
 units

log𝑦=𝑎+𝛽log𝑋

If 𝑋↑
 1%, then 𝑦↑
 by 𝛽
%

## Table of how an increase in x (binary) affects y

𝑦=𝑎+𝛽𝑋

𝑦
 is 𝛽
 units higher for cases when 𝑋=1
 than when 𝑋=0
.

log𝑦=𝑎+𝛽𝑋

𝑦
 is about 100∗𝛽
 % higher for cases when 𝑋=1
 than when 𝑋=0

## Table of how an increase in x (categorical) affects y
| X (original variable) | X1 | X2 |
|----------------------|----|----|
| 0                  | 0  | 0  |
| 1                  | 1  | 0  |
| 2                  | 0  | 1  |

Then

| If the model is...                         | \( a \) means                                              | \( \beta_1 \) means                                                    | \( \beta_2 \) means                                                    |
|--------------------------------------------|-----------------------------------------------------------|------------------------------------------------------------------------|------------------------------------------------------------------------|
| \( y = a + \beta_1 X_1 + \beta_2 X_2 \)    | The average value of \( y \) is \( a \) for group 0 (because \( X_1 = X_2 = 0 \) if \( X = 0 \)) | \( y \) is \( \beta_1 \) units higher on average when \( X = 1 \) than when \( X = 0 \). | \( y \) is \( \beta_2 \) units higher on average when \( X = 2 \) than when \( X = 0 \). |
| \( \log y = a + \beta_1 X_1 + \beta_2 X_2 \) | The average value of \( \log y \) is \( a \) for group 0 (because \( X_1 = X_2 = 0 \) if \( X = 0 \)) | \( y \) is about \( 100 \times \beta_1 \)% higher on average when \( X = 1 \) than when \( X = 0 \). | \( y \) is about \( 100 \times \beta_2 \)% higher on average when \( X = 2 \) than


Relationship of size on price: 1.53+0.18∗𝐼𝑑𝑒𝑎𝑙
- A 1% increase in size is associated with a 1.53% higher price for non ideal diamonds

- A 1% increase in size is associated with a 1.71% higher price for ideal diamonds

Relationship of cut on price: 0.33+0.18∗𝑙𝑜𝑔(𝑐𝑎𝑟𝑎𝑡)
- For 1 carat diamonds (𝑙𝑜𝑔(1)=0
), ideal diamonds are 33% more expensive than non-ideal diamonds

- For 2 carat diamonds (𝑙𝑜𝑔(2)=0.693
), ideal diamonds are 45% more expensive than non-ideal diamonds

How do you interpret 𝛽3=0.18
?

𝛽3≠0
- implies that the relationship between carat size and price is different for ideal and non-ideal diamonds.

- Mathematically: 1%↑
 in carat →
 price increases by 1.53% for non-ideal but 1.71% for ideal

- Graphically, the difference in the slope of those carat-price relationships for ideal/non-ideal diamonds is 𝛽3

- Economically, you might say that the value of a larger ring is even more valuable for better cut diamonds

𝛽3≠0
- implies that the relationship between cut quality and price is different for diamonds of different sizes.

- Mathematically: 1 carat diamonds that are ideal are 33% more expensive than non-ideal diamonds, but 2 carat ideal diamonds are 45% more expensive than non-ideal diamonds

- Graphically, the difference in the slope of those cut quality-price relationships for small/large diamonds is 𝛽3

- Economically, you might say that the value of a better cut is even more valuable for larger diamonds



In [1]:
import pandas as pd
from statsmodels.formula.api import ols as sm_ols
import numpy as np
import seaborn as sns
from statsmodels.iolib.summary2 import summary_col # nicer tables


In [2]:
url = 'https://github.com/LeDataSciFi/ledatascifi-2025/blob/main/data/Fannie_Mae_Plus_Data.gzip?raw=true'
fannie_mae = pd.read_csv(url,compression='gzip') 

## Clean the data and create variables you want

In [3]:
fannie_mae = (fannie_mae
                  # create variables
                  .assign(l_credscore = np.log(fannie_mae['Borrower_Credit_Score_at_Origination']),
                          l_LTV = np.log(fannie_mae['Original_LTV_(OLTV)']),
                          l_int = np.log(fannie_mae['Original_Interest_Rate']),
                          Origination_Date = lambda x: pd.to_datetime(x['Origination_Date']),
                          Origination_Year = lambda x: x['Origination_Date'].dt.year,
                          const = 1
                         )
                  .rename(columns={'Original_Interest_Rate':'int'}) # shorter name will help the table formatting
             )

# create a categorical credit bin var with "pd.cut()"
fannie_mae['creditbins']= pd.cut(fannie_mae['Co-borrower_credit_score_at_origination'],
                                 [0,579,669,739,799,850],
                                 labels=['Very Poor','Fair','Good','Very Good','Exceptional'])


## Statsmodels

As before, the psuedocode:
```python
model = sm_ols(<formula>, data=<dataframe>)
result=model.fit()

# you use result to print summary, get predicted values (.predict) or residuals (.resid)
```

Now, let's save each regression's result with a different name, and below this, output them all in one nice table:

In [4]:
# one var: 'y ~ x' means fit y = a + b*X

reg1 = sm_ols('int ~  Borrower_Credit_Score_at_Origination ', data=fannie_mae).fit()

reg1b= sm_ols('int ~  l_credscore  ',  data=fannie_mae).fit()

reg1c= sm_ols('l_int ~  Borrower_Credit_Score_at_Origination  ',  data=fannie_mae).fit()

reg1d= sm_ols('l_int ~  l_credscore  ',  data=fannie_mae).fit()

# multiple variables: just add them to the formula
# 'y ~ x1 + x2' means fit y = a + b*x1 + c*x2
reg2 = sm_ols('int ~  l_credscore + l_LTV ',  data=fannie_mae).fit()

# interaction terms: Just use *
# Note: always include each variable separately too! (not just x1*x2, but x1+x2+x1*x2)
reg3 = sm_ols('int ~  l_credscore + l_LTV + l_credscore*l_LTV',  data=fannie_mae).fit()
      
# categorical dummies: C() 
reg4 = sm_ols('int ~  C(creditbins)  ',  data=fannie_mae).fit()

reg5 = sm_ols('int ~  C(creditbins)  -1', data=fannie_mae).fit()


Ok, time to output them:

In [5]:
# now I'll format an output table
# I'd like to include extra info in the table (not just coefficients)
info_dict={'R-squared' : lambda x: f"{x.rsquared:.2f}",
           'Adj R-squared' : lambda x: f"{x.rsquared_adj:.2f}",
           'No. observations' : lambda x: f"{int(x.nobs):d}"}

# q4b1 and q4b2 name the dummies differently in the table, so this is a silly fix
reg4.model.exog_names[1:] = reg5.model.exog_names[1:]  

# This summary col function combines a bunch of regressions into one nice table
print('='*108)
print('                  y = interest rate if not specified, log(interest rate else)')
print(summary_col(results=[reg1,reg1b,reg1c,reg1d,reg2,reg3,reg4,reg5], # list the result obj here
                  float_format='%0.2f',
                  stars = True, # stars are easy way to see if anything is statistically significant
                  model_names=['1','2',' 3 (log)','4 (log)','5','6','7','8'], # these are bad names, lol. Usually, just use the y variable name
                  info_dict=info_dict,
                  regressor_order=[ 'Intercept','Borrower_Credit_Score_at_Origination','l_credscore','l_LTV','l_credscore:l_LTV',
                                  'C(creditbins)[Very Poor]','C(creditbins)[Fair]','C(creditbins)[Good]','C(creditbins)[Vrey Good]','C(creditbins)[Exceptional]']
                  )
     )

                  y = interest rate if not specified, log(interest rate else)

                                        1        2      3 (log) 4 (log)     5         6        7        8   
------------------------------------------------------------------------------------------------------------
Intercept                            11.58*** 45.37*** 2.87***  9.50***  44.13*** -16.81*** 6.65***         
                                     (0.05)   (0.29)   (0.01)   (0.06)   (0.30)   (4.11)    (0.08)          
Borrower_Credit_Score_at_Origination -0.01***          -0.00***                                             
                                     (0.00)            (0.00)                                               
l_credscore                                   -6.07***          -1.19*** -5.99*** 3.22***                   
                                              (0.04)            (0.01)   (0.04)   (0.62)                    
l_LTV                                            

# Today. Work in groups. Refer to the lectures. 

You might need to print out a few individual regressions with more decimals.

1. Interpret coefs in model 1-4
1. Interpret coefs in model 5
1. Interpret coefs in model 6 (and visually?)
1. Interpret coefs in model 7 (and visually? + comp to table)
1. Interpret coefs in model 8 (and visually? + comp to table)
1. Add l_LTV  to Model 8 and interpret (and visually?)





Model 1
 - beta_0: The interest rate is 11.58 if your credit score is 0
 - beta_1: For a 1 unit increase in credit score, the interest rate will fall by 0.86 b.p.
 - If your credit score = 700: 5.56 p.p
 - If your credit score = 707: 5.49 p.p
 - So, credit score up 7 points --> 6 b.p. drop in my interest rate

In [6]:
11.58 - .0086*(700)

5.5600000000000005

Model 2
 - beta_1: -6.07
 - beta_0: The interest rate is 45.37 if your credit score is 0
 - beta_1: For a 1%  increase in credit score, the interest rate will fall by 0.0607 p.p.
 - If your credit score = 700: 5.0649 p.p
 - If your credit score = 707: 5.5445 p.p
 - So, credit score up 7 points --> 6 b.p. drop in my interest rate


In [7]:
import numpy as np

45.37-6.07*(np.log(700))

5.604942366286529

Model 3
 - beta_0: 2.87
 - beta_1: -.0017
 - beta_0: The interest rate is 17.64 if your credit score is 0
 - beta_1: For a 1 unit increase in credit score, the interest rate will fall by 0.17 %.
 - If your credit score = 700: 5.3656 p.p
 - If your credit score = 707: 5.3021 p.p
 - So, credit score up 7 points --> 6 b.p. drop in my interest rate


In [8]:
np.exp(2.87)

17.63701819983732

In [9]:
np.exp(2.87-.0017*700)

5.365555971121975

In [10]:
np.exp(2.87-.0017*707)

5.302084260759588

Model 4
 - beta_0: 9.50
 - beta_1: -1.19
 - beta_0: The interest rate is 13359.73 if your credit score is 1

 - beta_1: For a 1 unit increase in credit score, the interest rate will fall by 0.17 %.
 - If your credit score = 700: 5.4971 p.p
 - If your credit score = 707: 5.4324 p.p
 - So, credit score up 7 points --> 6 b.p. drop in my interest rate

In [11]:
np.exp(9.5)

13359.726829661873

In [12]:
print(np.exp(9.5-1.19*np.log(700)))
print(np.exp(9.5-1.19*np.log(707)))

5.497065482961815
5.432359161191803


Model 5: int = 44.13 - 5.99 log(Cred) - 0.15 log(LTV)
 - beta_0: 44.13
 - beta_1: -5.99
 - beta_2: -0.15

 - beta_1: For a 1 unit increase in credit score, the interest rate will fall by 0.17 %.
 - If your credit score = 700: 5.5263 p.p
 - If your credit score = 707: 5.4667 p.p
 - So, credit score up 7 points --> 6 b.p. drop in my interest rate

In [22]:
print(fannie_mae[['Original_LTV_(OLTV)']].mean().iloc[0])

print(44.13 - 5.99*np.log(700) + .15*np.log(70))
print(44.13 - 5.99*np.log(707) + .15*np.log(70))

70.05729498363424
5.52630307939741
5.4667005975869305


Model 6: int = -16.81 + 3.22 log(Cred) +14.61log(LTV) - 2.18 log(Cred)*log(LTV)

 - If your credit score = 700: 6.3107 p.p
 - If your credit score = 707: 5.2506 p.p
 - So, credit score up 7 points --> 6 b.p. drop in my interest rate

In [20]:
print(-16.18 + 3.22*np.log(700) + 14.61*np.log(70)-2.18*np.log(700)*np.log(70))
print(-16.18 + 3.22*np.log(707) + 14.61*np.log(70)-2.18*np.log(707)*np.log(70))

6.310724843682131
6.250607734464772


Model 7 & 8: Look at regression, all there

## Vocab

For a given X1 coefficient in a regression

null hypothesis: Hypothesis assuming nothing is related/correlated

   - X1 is not related to y

std errors: estimates the std deviation of the coefficient

t stat:  = ceof/se

p-value: probabiility the non-zero beta is not zero by random chance

   - one star, p < 10%, two star, p < 5%, three star, p < 1%

economic significance: a reasonable change in x associated with a large change in y

 - reasonable = B caputures a one unit change which might be tiny or huge
 - Good Trick: scale cont vars by their STD so that a one unit change in X is a STD
 - large chg in y = compare coefficients to avg and std of y